text/0028-generalized-entity-data-source.md

```diff
@@ -0,0 +1,569 @@
+for future users, who will see two names for the same thing.
+- not expose `Value` and just use `EvalResult`, even in the new constructors.
+This provides maximum backwards compatibility, like the first option, and
+also avoids the possible confusion generated by the first option, but comes
+at a notable performance cost (due to the inefficient
+`EvalResult`-to-`Value` conversion required under the hood) and might
+generate a different kind of confusion as `EvalResult` is an awkward name in
+the context of these constructors.
+
+This RFC currently proposes we take the third option.
+
+Fourth and finally, we remove `Entities::evaluate()`, as all `Entities` are now
+stored in their evaluated form by default.
+This is not a breaking change because `Entities::evaluate()` has not yet been
+released in any Cedar release.
+
+## Drawbacks
+
+- This RFC represents some breaking changes for users upgrading to a new major
+version of Cedar. All breaking changes come with some cost in terms of user
+experience for existing users. This RFC contends that the benefit outweighs the
+cost in this case.
+
+- `EntityDataSource` represents a new contract we're making with users.  We
+could be locking ourselves in to some implementation choices that currently we'd
+be free to change behind the scenes. Similarly with exposing (even a wrapped
+version of) `Value`.
+
+- Some errors will occur at different times, or more lazily. In particular,
+today, if any entity attribute is invalid, that error will _always_ be caught
+during any call of `is_authorized()`. With this RFC, if an entity is never used
+during an evaluation, its attributes may never be evaluated and thus may not
+cause an error. Changing the behavior of some `is_authorized()` requests from
+"error" to "valid" is not technically a breaking change (otherwise, adding any
+new language syntax would be a breaking change), but this still might surprise
+some users.
+
+- The current proposed definition of `EntityDataSource` assumes that every
+implementation of the trait handles transitive closure properly, and has no way
+to check this assumption. If users make mistakes in this area, they may silently
+get the wrong authorization results.
+(This is also true of implementation errors such as loading the wrong attribute
+values, etc; but errors with TC may feel more insidious somehow.)
+
+- A user could write an `EntityDataSource` implementation which represents an
+infinitely large entity data store -- e.g., which always responds `true` to
+every possible UID in `entity_exists()`. This may be a theoretical issue for
+validation?
+
+## Alternatives
+
+There are a lot of branching alternatives to all or parts of this RFC; this is
+not a complete discussion of the decision tree.
+
+Some of these alternatives are mutually exclusive with each other, and some are
+not.
+
+### Alternative A
+
+We could do only the first component of the RFC, at least for now.
```

Comment:
The other changes are in large part geared towards helping users write `EntityDataSource` implementations easily and performantly.  It would be a shame to provide users only a more-confusing and less-efficient interface, have it be widely adopted, and only later decide to release improvements that require migration in order to take advantage of.

Reply:
You should write this in the RFC. :)

Reply:
Added more discussion to Alternative A